File: mobile/test/features/channels/agent_turn_receipt_test.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:buzz/features/channels/agent_turn_receipt.dart';
import 'package:buzz/features/channels/agent_turn_receipt_footer.dart';
import 'package:buzz/features/channels/timeline_message.dart';
import 'package:buzz/shared/relay/relay.dart';
import 'package:buzz/shared/theme/theme.dart';

import '../../helpers/widget_helpers.dart';

const _channelId = 'ch1';
const _agent = 'a11ce';
const _stranger = 'deadbeef';

NostrEvent _message({
  required String id,
  String pubkey = _agent,
  String content = 'done',
  int createdAt = 1000,
}) => NostrEvent(
  id: id,
  pubkey: pubkey,
  createdAt: createdAt,
  kind: EventKind.streamMessage,
  tags: const [
    ['h', _channelId],
  ],
  content: content,
  sig: '',
);

/// A well-formed NIP-AR receipt. [bodyModel] defaults to [model] so the tag and
/// the body agree; passing it explicitly is how a disagreement is staged.
NostrEvent _receipt({
  String id = 'receipt',
  required List<String> messageIds,
  String pubkey = _agent,
  int createdAt = 1100,
  String model = 'claude-opus-4-5',
  String? bodyModel,
  String harness = 'claude-agent-acp',
  Map<String, Object?> turn = const {
    'inputTokens': 191261,
    'outputTokens': 683,
    'cacheReadTokens': 122407,
    'cacheWriteTokens': 4096,
  },
  List<List<String>>? eTags,
}) => NostrEvent(
  id: id,
  pubkey: pubkey,
  createdAt: createdAt,
  kind: EventKind.agentTurnReceipt,
  tags: [
    const ['h', _channelId],
    ...(eTags ??
        [
          for (final messageId in messageIds) ['e', messageId],
        ]),
    ['model', model],
  ],
  content: jsonEncode({
    'model': bodyModel ?? model,
    'harness': harness,
    'turn': turn,
  }),
  sig: '',
);

AgentTurnReceipt _parse(NostrEvent event) => AgentTurnReceipt.fromEvent(event)!;

void main() {
  group('turn receipts attach to the timeline', () {
    test('a receipt authored by someone other than the message author is '
        'ignored', () {
      final messages = formatTimeline([
        _message(id: 'm1'),
        _receipt(messageIds: const ['m1'], pubkey: _stranger),
      ]);

      expect(messages.single.turnReceipt, isNull);
    });

    test('a receipt authored by the message author is attached', () {
      final messages = formatTimeline([
        _message(id: 'm1'),
        _receipt(messageIds: const ['m1']),
      ]);

      expect(messages.single.turnReceipt?.model, 'claude-opus-4-5');
    });

    test('a turn that published three messages attaches its receipt once, to '
        'the last of them', () {
      final messages = formatTimeline([
        _message(id: 'm1', createdAt: 1000),
        _message(id: 'm2', createdAt: 1001),
        _message(id: 'm3', createdAt: 1002),
        _receipt(messageIds: const ['m1', 'm2', 'm3']),
      ]);

      expect(
        messages.map((message) => message.turnReceipt != null),
        [false, false, true],
        reason: 'one turn is one footer, under the turn\'s final message',
      );
    });

    test('a receipt anchors to the last named message this client actually '
        'holds', () {
      // The turn published three messages; this client never loaded the third.
      final messages = formatTimeline([
        _message(id: 'm1', createdAt: 1000),
        _message(id: 'm2', createdAt: 1001),
        _receipt(messageIds: const ['m1', 'm2', 'm3']),
      ]);

      expect(messages.map((message) => message.turnReceipt != null), [
        false,
        true,
      ]);
    });

    test('a receipt naming no message this client holds is never rendered', () {
      final messages = formatTimeline([
        _message(id: 'm1'),
        _receipt(messageIds: const ['somewhere-else']),
      ]);

      expect(messages.single.turnReceipt, isNull);
    });

    test('a forged receipt is dropped rather than walked back to an earlier '
        'message it names', () {
      // The attacker names their own message first and the victim's last, so a
      // consumer that fell back on a failed trust check would still render
      // attacker-supplied numbers — under the attacker's own message, but
      // sourced from a receipt that lies about the victim's.
      final messages = formatTimeline([
        _message(id: 'mine', pubkey: _stranger, createdAt: 1000),
        _message(id: 'theirs', pubkey: _agent, createdAt: 1001),
        _receipt(messageIds: const ['mine', 'theirs'], pubkey: _stranger),
      ]);

      expect(messages.map((message) => message.turnReceipt), [isNull, isNull]);
    });

    test('a receipt whose model tag disagrees with its content is ignored', () {
      final messages = formatTimeline([
        _message(id: 'm1'),
        _receipt(
          messageIds: const ['m1'],
          model: 'claude-opus-4-5',
          bodyModel: 'something-cheaper',
        ),
      ]);

      expect(messages.single.turnReceipt, isNull);
    });

    test('a receipt whose only e tag carries a NIP-10 marker is ignored', () {
      // A marked `e` tag is a thread link, not an annotation target. A receipt
      // is an overlay on the messages it names, never a reply to them.
      final messages = formatTimeline([
        _message(id: 'm1'),
        _receipt(
          messageIds: const ['m1'],
          eTags: const [
            ['e', 'm1', '', 'reply'],
          ],
        ),
      ]);

      expect(messages.single.turnReceipt, isNull);
    });

    test('the newer receipt wins when two of them anchor to the same '
        'message', () {
      final messages = formatTimeline([
        _message(id: 'm1'),
        _receipt(
          id: 'older',
          messageIds: const ['m1'],
          createdAt: 1100,
          model: 'first-model',
        ),
        _receipt(
          id: 'newer',
          messageIds: const ['m1'],
          createdAt: 1200,
          model: 'second-model',
        ),
      ]);

      expect(messages.single.turnReceipt?.model, 'second-model');
    });

    test('a deleted receipt is ignored', () {
      final messages = formatTimeline([
        _message(id: 'm1'),
        _receipt(id: 'receipt', messageIds: const ['m1']),
        NostrEvent(
          id: 'del',
          pubkey: _agent,
          createdAt: 1200,
          kind: EventKind.deletion,
          tags: const [
            ['e', 'receipt'],
          ],
          content: '',
          sig: '',
        ),
      ]);

      expect(messages.single.turnReceipt, isNull);
    });

    test('kind 44201 is an overlay kind, not a timeline row or a message', () {
      // A receipt has no row of its own: counting it as channel content would
      // create a phantom unread for an event no reader can see.
      expect(
        EventKind.channelEventKinds,
        contains(EventKind.agentTurnReceipt),
        reason: 'the channel subscription must not discard receipts',
      );
      expect(
        EventKind.channelAuxEventKinds,
        contains(EventKind.agentTurnReceipt),
      );
      expect(
        EventKind.channelMessageEventKinds,
        isNot(contains(EventKind.agentTurnReceipt)),
      );
      expect(
        EventKind.channelTimelineContentKinds,
        isNot(contains(EventKind.agentTurnReceipt)),
      );

      // And it renders no row of its own.
      final messages = formatTimeline([
        _message(id: 'm1'),
        _receipt(messageIds: const ['m1']),
      ]);
      expect(messages, hasLength(1));
    });
  });

  group('unreported counts are never rendered as zero', () {
    test('a null token count parses as not reported, not as zero', () {
      final receipt = _parse(
        _receipt(
          messageIds: const ['m1'],
          turn: const {
            'inputTokens': null,
            'outputTokens': 683,
            'cacheReadTokens': null,
          },
        ),
      );

      expect(receipt.inputTokens.state, ReceiptCountState.notReported);
      expect(receipt.inputTokens.value, isNull);
      // The cache fields may be omitted entirely; absence reads the same.
      expect(receipt.cacheWriteTokens.state, ReceiptCountState.notReported);
      expect(receipt.outputTokens.value, 683);
    });

    test('a reported zero parses as a real zero', () {
      final receipt = _parse(
        _receipt(messageIds: const ['m1'], turn: const {'outputTokens': 0}),
      );

      expect(receipt.outputTokens.state, ReceiptCountState.reported);
      expect(receipt.outputTokens.display, '0');
    });

    test('a token count that is not a non-negative integer reads as unknown, '
        'not as zero', () {
      final receipt = _parse(
        _receipt(
          messageIds: const ['m1'],
          turn: const {'inputTokens': -5, 'outputTokens': 'lots'},
        ),
      );

      expect(receipt.inputTokens.state, ReceiptCountState.unreadable);
      expect(receipt.outputTokens.state, ReceiptCountState.unreadable);
      expect(receipt.inputTokens.display, isNot('0'));
      expect(receipt.outputTokens.display, isNot('0'));
    });
  });

  group('AgentTurnReceiptFooter', () {
    Future<void> pumpFooter(WidgetTester tester, AgentTurnReceipt receipt) {
      return tester.pumpWidget(
        WidgetHelpers.testable(child: AgentTurnReceiptFooter(receipt: receipt)),
      );
    }

    testWidgets('names the model the turn ran on', (tester) async {
      await pumpFooter(tester, _parse(_receipt(messageIds: const ['m1'])));

      expect(find.text('claude-opus-4-5'), findsOneWidget);
    });

    testWidgets('renders every reported token count', (tester) async {
      await pumpFooter(tester, _parse(_receipt(messageIds: const ['m1'])));

      expect(find.text('· in 191,261'), findsOneWidget);
      expect(find.text('· out 683'), findsOneWidget);
      expect(find.text('· cache read 122,407'), findsOneWidget);
      expect(find.text('· cache write 4,096'), findsOneWidget);
    });

    testWidgets('renders a count the harness did not report as a dash, never '
        'as zero', (tester) async {
      await pumpFooter(
        tester,
        _parse(
          _receipt(
            messageIds: const ['m1'],
            turn: const {'inputTokens': 12, 'outputTokens': null},
          ),
        ),
      );

      expect(find.text('· out —'), findsOneWidget);
      expect(find.text('· out 0'), findsNothing);
      expect(find.text('· cache read —'), findsOneWidget);
      expect(find.text('· cache write —'), findsOneWidget);
      // Falsifiability: the dash is not simply what every slot renders.
      expect(find.text('· in 12'), findsOneWidget);
    });

    testWidgets('renders a reported zero as zero', (tester) async {
      await pumpFooter(
        tester,
        _parse(
          _receipt(messageIds: const ['m1'], turn: const {'outputTokens': 0}),
        ),
      );

      expect(find.text('· out 0'), findsOneWidget);
      expect(find.text('· out —'), findsNothing);
    });

    testWidgets('speaks one sentence instead of its glyphs', (tester) async {
      final handle = tester.ensureSemantics();
      await pumpFooter(
        tester,
        _parse(
          _receipt(
            messageIds: const ['m1'],
            turn: const {'inputTokens': 10, 'outputTokens': null},
          ),
        ),
      );

      expect(
        find.bySemanticsLabel(
          'Agent turn ran on claude-opus-4-5. Tokens: input 10, '
          'output not reported, cache read not reported, cache write not '
          'reported.',
        ),
        findsOneWidget,
      );
      // The dash glyph must not be what a screen reader announces.
      expect(find.bySemanticsLabel('· out —'), findsNothing);
      handle.dispose();
    });

    testWidgets('lines up under the message body, like the reaction row', (
      tester,
    ) async {
      await pumpFooter(tester, _parse(_receipt(messageIds: const ['m1'])));

      // The indent lives in the widget, not at its two call sites, which is
      // what keeps the channel and thread views from drifting apart.
      final padding = tester.widget<Padding>(
        find
            .ancestor(of: find.byType(Wrap), matching: find.byType(Padding))
            .first,
      );
      expect(
        (padding.padding as EdgeInsets).left,
        messageAvatarSize + messageAvatarContentGap,
      );
    });
  });
}

File: desktop/src/features/messages/lib/agentTurnReceipt.test.mjs
/**
 * NIP-AR receipt parsing, trust, and anchor selection.
 *
 * The two rules under test are the ones the NIP calls mandatory, and both
 * fail silently if broken: a forged receipt renders attacker-supplied numbers
 * under someone else's name, and a per-message render multiplies one turn's
 * spend by however many messages it happened to split into.
 */

import assert from "node:assert/strict";
import test from "node:test";

import {
  isNewerTurnReceipt,
  parseTurnReceiptEvent,
  selectTurnReceiptAnchor,
} from "./agentTurnReceipt.ts";
import { KIND_AGENT_TURN_RECEIPT } from "@/shared/constants/kinds";

const CHANNEL_ID = "36411e44-0e2d-4cfe-bd6e-567eb169db9f";
const AGENT_PUBKEY = "a".repeat(64);
const STRANGER_PUBKEY = "f".repeat(64);
const MSG_1 = "1".repeat(64);
const MSG_2 = "2".repeat(64);
const MSG_3 = "3".repeat(64);

function receiptEvent(overrides = {}) {
  const {
    targets = [MSG_1],
    model = "claude-opus-4-5",
    modelTag = model,
    turn = { inputTokens: 191261, outputTokens: 683, cacheReadTokens: 122407 },
    harness = "claude-agent-acp",
    content,
    ...rest
  } = overrides;
  return {
    id: "9".repeat(64),
    pubkey: AGENT_PUBKEY,
    kind: KIND_AGENT_TURN_RECEIPT,
    created_at: 1_700_000_100,
    content: content ?? JSON.stringify({ model, harness, turn }),
    tags: [
      ["h", CHANNEL_ID],
      ...targets.map((id) => ["e", id]),
      ["model", modelTag],
    ],
    sig: "sig",
    ...rest,
  };
}

function message(id, pubkey = AGENT_PUBKEY) {
  return {
    id,
    pubkey,
    kind: 9,
    created_at: 1_700_000_000,
    content: "hi",
    tags: [["h", CHANNEL_ID]],
    sig: "sig",
  };
}

function anchorFor(event, held) {
  const parsed = parseTurnReceiptEvent(event);
  if (!parsed) return null;
  const byId = new Map(held.map((m) => [m.id, m]));
  return selectTurnReceiptAnchor({
    parsed,
    receiptPubkey: event.pubkey,
    getHeldEvent: (id) => byId.get(id),
    resolveAuthor: (target) => target.pubkey,
  });
}

test("a receipt carries the model, harness, and every reported count", () => {
  const parsed = parseTurnReceiptEvent(receiptEvent());
  assert.equal(parsed.receipt.model, "claude-opus-4-5");
  assert.equal(parsed.receipt.harness, "claude-agent-acp");
  assert.equal(parsed.receipt.inputTokens, 191261);
  assert.equal(parsed.receipt.outputTokens, 683);
  assert.equal(parsed.receipt.cacheReadTokens, 122407);
  assert.deepEqual(parsed.targetIds, [MSG_1]);
});

test("a count the harness did not report parses as null, never as zero", () => {
  const parsed = parseTurnReceiptEvent(
    receiptEvent({ turn: { inputTokens: 10, outputTokens: null } }),
  );
  assert.equal(parsed.receipt.inputTokens, 10);
  assert.equal(parsed.receipt.outputTokens, null);
  // Omitted entirely (the NIP's shape for unobservable cache counters).
  assert.equal(parsed.receipt.cacheReadTokens, null);
  assert.equal(parsed.receipt.cacheWriteTokens, null);
});

test("a genuine zero count survives parsing as zero", () => {
  const parsed = parseTurnReceiptEvent(
    receiptEvent({ turn: { inputTokens: 0, cacheWriteTokens: 0 } }),
  );
  assert.equal(parsed.receipt.inputTokens, 0);
  assert.equal(parsed.receipt.cacheWriteTokens, 0);
});

test("a count that is not a non-negative integer is treated as unreported", () => {
  for (const bad of ["1000", -5, 1.5, Number.NaN, Number.POSITIVE_INFINITY]) {
    const parsed = parseTurnReceiptEvent(
      receiptEvent({ turn: { inputTokens: bad } }),
    );
    assert.equal(parsed.receipt.inputTokens, null, `${String(bad)}`);
  }
});

test("a receipt with no e tags, no model, or unparseable content is rejected", () => {
  assert.equal(parseTurnReceiptEvent(receiptEvent({ targets: [] })), null);
  assert.equal(parseTurnReceiptEvent(receiptEvent({ model: "  " })), null);
  assert.equal(parseTurnReceiptEvent(receiptEvent({ harness: "" })), null);
  assert.equal(
    parseTurnReceiptEvent(receiptEvent({ content: "not json" })),
    null,
  );
  assert.equal(parseTurnReceiptEvent({ ...receiptEvent(), kind: 9 }), null);
});

test("a receipt whose model tag disagrees with its body is rejected", () => {
  assert.equal(
    parseTurnReceiptEvent(
      receiptEvent({ model: "claude-opus-4-5", modelTag: "claude-haiku-4-5" }),
    ),
    null,
  );
});

test("a receipt whose e tag carries a NIP-10 marker is rejected", () => {
  // The predicate that channel-scopes 44201 is the same one that resolves
  // thread ancestry, so a marked tag would make a receipt a reply and inflate
  // the annotated message's reply_count. The relay refuses it at ingest; both
  // clients refuse it too, so neither depends on that and neither drifts.
  const marked = receiptEvent();
  marked.tags = marked.tags.map((tag) =>
    tag[0] === "e" ? [tag[0], tag[1], "", "reply"] : tag,
  );
  assert.equal(parseTurnReceiptEvent(marked), null);

  // An unmarked tag with a relay hint is still a plain reference.
  const hinted = receiptEvent();
  hinted.tags = hinted.tags.map((tag) =>
    tag[0] === "e" ? [tag[0], tag[1], "wss://relay.example"] : tag,
  );
  assert.notEqual(parseTurnReceiptEvent(hinted), null);
});

test("a receipt from a different pubkey than the message author is ignored", () => {
  const forged = receiptEvent({ pubkey: STRANGER_PUBKEY, targets: [MSG_1] });
  assert.equal(anchorFor(forged, [message(MSG_1, AGENT_PUBKEY)]), null);
});

test("a three-message turn anchors once, under the last message held", () => {
  const event = receiptEvent({ targets: [MSG_1, MSG_2, MSG_3] });
  const held = [message(MSG_1), message(MSG_2), message(MSG_3)];
  assert.equal(anchorFor(event, held), MSG_3);
});

test("a turn whose last messages are missing anchors to the last one held", () => {
  const event = receiptEvent({ targets: [MSG_1, MSG_2, MSG_3] });
  assert.equal(anchorFor(event, [message(MSG_1), message(MSG_2)]), MSG_2);
});

test("a receipt naming no message this client holds anchors nowhere", () => {
  const event = receiptEvent({ targets: [MSG_1, MSG_2] });
  assert.equal(anchorFor(event, []), null);
});

test("duplicate e tags do not change which message a receipt anchors to", () => {
  const event = receiptEvent({ targets: [MSG_1, MSG_2, MSG_1] });
  const parsed = parseTurnReceiptEvent(event);
  assert.deepEqual(parsed.targetIds, [MSG_1, MSG_2]);
  assert.equal(anchorFor(event, [message(MSG_1), message(MSG_2)]), MSG_2);
});

test("the newer of two receipts wins the same anchor, order-independently", () => {
  const older = parseTurnReceiptEvent(receiptEvent({ created_at: 10 }));
  const newer = parseTurnReceiptEvent(
    receiptEvent({ created_at: 20, id: "8".repeat(64) }),
  );
  assert.equal(isNewerTurnReceipt(newer, older), true);
  assert.equal(isNewerTurnReceipt(older, newer), false);
  // Same second: the id breaks the tie both ways, so delivery order cannot
  // decide which receipt is displayed.
  const tieA = parseTurnReceiptEvent(
    receiptEvent({ created_at: 10, id: "7".repeat(64) }),
  );
  const tieB = parseTurnReceiptEvent(
    receiptEvent({ created_at: 10, id: "8".repeat(64) }),
  );
  assert.equal(isNewerTurnReceipt(tieB, tieA), true);
  assert.equal(isNewerTurnReceipt(tieA, tieB), false);
});
